### ISSUE TITLE
cannot mount zpool: directory is not empty

### ISSUE THREAD

**Issue by c0xc** [openzfs/zfs#9207]
### System information

Type | Version/Name
 --- | --- 
Distribution Name	| Fedora
Distribution Version	| 29
Linux Kernel	| Linux 5.2.7-100.fc29.x86_64
Architecture	| x86_64
ZFS Version	| 0.8.1-1
SPL Version	| 0.8.1-1

### Describe the problem you're observing

The zpool is not mounted on startup. Trying to mount it manually fails:

> cannot mount '/z-main': directory is not empty

Indeed, that directory (which was created automatically) is not empty because it contains empty subdirectories, one for each ZFS volume in that pool. It seems like those mountpoints are preventing the pool itself from being mounted. I feel like that shouldn't happen (if those mountpoints are created automatically, they should be removed automatically as well).

### Describe how to reproduce the problem

Whenever the system is rebooted, the pool isn't available. Mounting it manually without removing those empty directories first fails. As it's a common scheme to use bind mounts in order to export the ZFS volumes via NFS, those have to be fixed as well.

I wrote a [script](https://github.com/c0xc/zfs-mount-check/blob/master/zpool-mount) that does all that and can be run on startup: It tries to mount the zpool (if not mounted already) and if that fails, it removes all empty subdirectories within the base mountpoint of the pool (e.g., `/z-main`; specified at the top of the script) and tries to mount it again. After that, it tries to fix the bind mounts (if specified).

Obviously, such a script shouldn't be necessary but it's a workaround for the moment and maybe it helps someone out there. I have found a similar bug, which was about a systemd service not being enabled, but zfs-import.target is enabled, so I don't think that's it. I vaguely remember a similar situation on a Debian system with systemd but in that case, it was sufficient to enable and restart zfs-mount.service.

**Comment by GregorKopka** [comment]
You can work around that by rm'ing the mountpoints (which might lead to unintended data loss in case theres something in there for some reason), or you can set `overlay=on` on the problematic filesystems to instruct ZFS to skip the empty check and mount regardless of an existing mountpoint being empty or not.

Note though that `overlay` isn't what some might expect when hearing that word (like functionality as https://en.wikipedia.org/wiki/Aufs or https://en.wikipedia.org/wiki/OverlayFS supply) as sadly someone, for some unknown reason, decided against `noemptycheckonmount` (which would have been the precise description of the function) as the name for this dataset property.

**Comment by c0xc** [comment]
Of course, I would never run `rm` on a mountpoint (only `rmdir`), especially without checking if it's an active mountpoint first. That would be suicidal. Please look at my makeshift script:
https://github.com/c0xc/zfs-mount-check/blob/master/zpool-mount

Also, I don't feel like using the `overlay` option is the right way to go because it's like covering up the issue rather than fixing it. Why are there even empty subdirectories in /${zpool} (which itself isn't mounted yet!)?

**Comment by h1z1** [comment]
Hit this all the time, it's infuriating.  The directory technically isn't empty (. and ..).    Can't speak for the history behind it but the code is I believe backward.  stat(2) and readdir are used depending on if the filesystem is zfs or not.   They are different logic though, I don't think stat checks file count, readdir does.  

```
        if ((flags & MS_OVERLAY) == 0 && !remount &&
            !dir_is_empty(mountpoint)) {
                zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
                    "directory is not empty"));
                return (zfs_error_fmt(hdl, EZFS_MOUNTFAILED,
                    dgettext(TEXT_DOMAIN, "cannot mount '%s'"), mountpoint));
        }
```

```
       if (stat(dirname, &st) < 0 || !S_ISDIR(st.st_mode)) {
                return (B_TRUE);
        }
```
```

        /*
         * If the statvfs call fails or the filesystem is not a ZFS
         * filesystem, fall back to the slow path which uses readdir.
         */
        if ((statfs64(dirname, &st) != 0) ||
            (st.f_type != ZFS_SUPER_MAGIC)) {
                return (dir_is_empty_readdir(dirname));
        }

```

**Comment by ttelford** [comment]
I feel your pain. I have dozens of mount points which have this error upon boot.

I only use ZFS -- everything on my system is a zfs mount, however, for whatever reason, `zfs mount -a` does not mount the filesystems in the correct order.

I'm not sure if it's because there is an issue with the mounting code, or if it's a case of `systemd` starting processes before zfs has a chance to fnish mountin an associated filesystem. For example, `/var/spool`, `/var/tmp`, `/var/log`, `/var/cache` all refuse to mount, because the fs isn't empty.

The only way I can see that happening is that a process (say, `rsyslog`) is starting up after zfs mounts `/var`, but before it can mount `/var/log.

**Comment by dswartz** [comment]
On 2020-01-30 16:23, Troy Telford wrote:
> I feel your pain. I have dozens of mount points which have this error
> upon boot.
> 
> I only use ZFS -- everything on my system is a zfs mount, however, for
> whatever reason, zfs mount -a does not mount the filesystems in the
> correct order.
> 
> I'm not sure if it's because there is an issue with the mounting code,
> or if it's a case of systemd starting processes before zfs has a
> chance to fnish mountin an associated filesystem. For example,
> /var/spool, /var/tmp, /var/log, /var/cache all refuse to mount,
> because the fs isn't empty.
> 
> The only way I can see that happening is that a process (say, rsyslog)
> is starting up after zfs mounts /var, but before it can mount
> `/var/log.

Back when I was using proxmox, I hit this all the time.  Proxmox 
automatically creates certain sub-directories in certain places, and if 
ZFS didn't mount the pool quickly enough, proxmox would create the 
sub-directories, preventing ZFS from mounting those filesystems.  It's 
been 3 years or so, so I have no idea if that problem still exists.

**Comment by Fabian-Gruenbichler** [comment]
On January 30, 2020 11:25 pm, dswartz wrote:
> On 2020-01-30 16:23, Troy Telford wrote:
>> I feel your pain. I have dozens of mount points which have this error
>> upon boot.
>> 
>> I only use ZFS -- everything on my system is a zfs mount, however, for
>> whatever reason, zfs mount -a does not mount the filesystems in the
>> correct order.
>> 
>> I'm not sure if it's because there is an issue with the mounting code,
>> or if it's a case of systemd starting processes before zfs has a
>> chance to fnish mountin an associated filesystem. For example,
>> /var/spool, /var/tmp, /var/log, /var/cache all refuse to mount,
>> because the fs isn't empty.
>> 
>> The only way I can see that happening is that a process (say, rsyslog)
>> is starting up after zfs mounts /var, but before it can mount
>> `/var/log.
> 
> Back when I was using proxmox, I hit this all the time.  Proxmox 
> automatically creates certain sub-directories in certain places, and if 
> ZFS didn't mount the pool quickly enough, proxmox would create the 
> sub-directories, preventing ZFS from mounting those filesystems.  It's 
> been 3 years or so, so I have no idea if that problem still exists.

if you don't tell proxmox that a directory you configured as storage is 
a mountpoint, it does not know. if you tell it, it will only treat it 
as usable and write to it if something is mounted there.

**Comment by c0xc** [comment]
It hasn't happened last time I had to reboot. Since I couldn't find any indication (like a comment in the [release notes](https://github.com/openzfs/zfs/releases)) that it's been identified and fixed, I'll keep my [zpool-mount script](https://github.com/c0xc/zfs-mount-check/blob/master/zpool-mount) for now. I hope I can remove it soon.

This is what the script output would typically look like when run automatically on startup:
```
ZPOOL NOT MOUNTED - /z-main/Share not a mountpoint
ATTEMPTING TO MOUNT ZFS...
ZFS complains about its mountpoint not being empty or something
rmdir: removing directory, '/z-main/Share'
rmdir: removing directory, '/z-main/'
ZFS MOUNTED
Attempting to re-mount bind mounts in /data...
```

As for the bind mounts (`/data/...` in this example, pointing to the zfs mountpoints): Those have been working fine for years until recently. The last time the system was rebooted, the filesystems weren't available causing other services like NFS and virtualization to fail. To be more precise, the zfs filesystems were mounted but the bind mounts were empty, i.e., they were mounted before the zfs filesystems they point to was available. It's recommended to try the generator but the script failed somehow so I've now added the mount option "x-systemd.after=zfs-mount.service" to make the system wait for zfs before bind-mounting it:
```
/z-... /data/... none defaults,bind,x-systemd.after=zfs-mount.service 0 0
```
